cmake/variables.cmake

```diff
@@ -153,13 +153,33 @@ if(CMAKE_SYSTEM_NAME STREQUAL "Windows")
   list(APPEND ML_COMPILE_DEFINITIONS BOOST_ALL_NO_LIB)
 endif()
 
-if(UNIX AND "$ENV{ML_DEBUG}" AND "$ENV{ML_COVERAGE}")
+# Perform a "Release" build by default...
+if(NOT CMAKE_BUILD_TYPE)
+  set(CMAKE_BUILD_TYPE Release)
+endif()
+
+# However, to keep in step with our historical
+# build system if ML_DEBUG is set to a boolean
+# true value, e.g. ML_DEBUG=1, then override
+# the build type to be Debug
+if("$ENV{ML_DEBUG}")
+  set(CMAKE_BUILD_TYPE Debug)
+endif()
+
+message(STATUS "CMAKE_BUILD_TYPE: ${CMAKE_BUILD_TYPE}")
+
+if(UNIX AND CMAKE_BUILD_TYPE STREQUAL Debug AND DEFINED ENV{ML_COVERAGE})
   set(ML_COVERAGE "--coverage")
 endif()
 
-if(NOT "$ENV{ML_DEBUG}")
+if(CMAKE_BUILD_TYPE STREQUAL Debug)
+  message(STATUS "IN DEBUG MODE")
+else()
+  message(STATUS "NOT IN DEBUG MODE")
+  message(STATUS "Adding NDEBUG and EXCLUDE_TRACE_LOGGING to ML_COMPILE_DEFINITIONS")
   list(APPEND ML_COMPILE_DEFINITIONS "NDEBUG" "EXCLUDE_TRACE_LOGGING")
   if(UNIX)
+    message(STATUS "Adding -O3 to OPTCFLAGS")
```

Comment:
`-O3` is showing up twice on the compilation lines.  I think it's because we're appending `OPTCFLAGS` onto the final compilation flags in two places:

```
bash-3.2$ find . -name '*.cmake' -type f | xargs grep OPTCFLAGS
./cmake/compiler/clang.cmake:  ${OPTCFLAGS}
./cmake/compiler/msvc.cmake:  ${OPTCFLAGS}
./cmake/compiler/gcc.cmake:  ${OPTCFLAGS}
./cmake/variables.cmake:    list(APPEND OPTCFLAGS "-O3")
./cmake/variables.cmake:      list(APPEND OPTCFLAGS "-Wdisabled-optimization")
./cmake/variables.cmake:    set(OPTCFLAGS $<$<CONFIG:Release>:/O2> "/Qfast_transcendentals" "/Qvec-report:1")
./cmake/variables.cmake:list(APPEND ML_CXX_FLAGS ${OPTCFLAGS})
```

Since CMake can append to lists, maybe we don't even need an `OPTCFLAGS` variable now?  Or even if we do still want to separate out the optimization flags we should only append them once.

Maybe it would be cleaner to append all the compiler-dependent arguments in the `compiler/*.cmake` files using the `$<$<CONFIG:Release>:-O3>` syntax to select the ones appropriate to the current build type?

Reply:
Also, related to this, `msvc.cmake` references a variable `${OPTCPPFLAGS}` that I don't think is set anywhere.

We seem to have lost some Eigen-related Windows optimization flags from https://github.com/elastic/ml-cpp/blob/c86b7174f20c427c0ce3b19f6a280edbaefe7773/mk/windows.mk#L102

The ones that I cannot see any more are:

> `-DEIGEN_VECTORIZE_SSE3 -DEIGEN_VECTORIZE_SSE4_1 -DEIGEN_VECTORIZE_SSE4_2`

Reply:
If it's cleaner it wouldn't hurt to define `-DEIGEN_VECTORIZE_SSE3 -DEIGEN_VECTORIZE_SSE4_1 -DEIGEN_VECTORIZE_SSE4_2` on all platforms. We want to use these instructions on all platforms. With clang and gcc Eigen can define them itself based on the compiler flags. With Visual C++ it cannot. But it wouldn't hurt for us to define these ourselves always.